File: ICESat2/ATL15_AA_0314_01km_002_02.nc.py
"""
This example code illustrates how to access and visualize an NSIDC
ATL15 vesion 2 netCDF-4 file in Python.

If you have any questions, suggestions, or comments on this example, please use
the HDF-EOS Forum (http://hdfeos.org/forums).  If you would like to see an
example of any other NASA HDF/HDF-EOS data product that is not listed in the
HDF-EOS Comprehensive Examples page (http://hdfeos.org/zoo), feel free to
contact us at eoshelp@hdfgroup.org or post it at the HDF-EOS Forum
(http://hdfeos.org/forums).

Usage:  save this script and run

    $python ATL15_AA_0314_01km_002_02.nc.py

The HDF file must be in your current working directory.

Tested under: Python 3.9.13 :: Miniconda
Last updated: 2023-05-05

"""

import os
import re
import h5py
import pyproj

import matplotlib as mpl
import matplotlib.pyplot as plt
import numpy as np

from mpl_toolkits.basemap import Basemap
from pyproj import Transformer

FILE_NAME = 'ATL15_AA_0314_01km_002_02.nc'
DATAFIELD_NAME = '/dhdt_lag1/dhdt'

with h5py.File(FILE_NAME, mode='r') as f:
    
    # Read dataset.
    datavar = f[DATAFIELD_NAME]
    
    # Subset data at time = 0.
    data = datavar[0,:,:]    
    _FillValue = datavar.attrs['_FillValue']
    units = datavar.attrs['units']
    long_name= datavar.attrs['long_name']
    
    # Handle fill value.
    data[data == _FillValue] = np.nan
    data = np.ma.masked_where(np.isnan(data), data)    

    # Read lat/lon.
    latvar = f['/dhdt_lag1/y']
    lat = latvar[:]
    
    lonvar = f['/dhdt_lag1/x']
    lon = lonvar[:]
    
    # Read time.
    timevar = f['/dhdt_lag1/time']
    time = timevar[:] 
    t_units = timevar.attrs['units']
    
    # Define the source and destination projections.
    src_proj = pyproj.CRS("EPSG:3031")
    dst_proj = pyproj.CRS("EPSG:4326")
    xv, yv = np.meshgrid(lon, lat)

    # Convert the coordinates.
    t = Transformer.from_crs(src_proj, dst_proj, always_xy=True)
    x, y = t.transform(xv, yv)

    m = Basemap(projection='spstere', resolution='l', boundinglat=-45,
                lon_0 = 0)
    m.drawcoastlines(linewidth=0.5)
    m.drawparallels(np.arange(-80, 0, 20), labels=[1, 0, 0, 0])
    m.drawmeridians(np.arange(-180, 181, 30), labels=[0, 0, 0, 1])
    m.pcolormesh(x, y, data, latlon=True)
    cb = m.colorbar()
    cb.set_label(units.decode('ascii', 'replace'))

    basename = os.path.basename(FILE_NAME)
    desc = long_name.decode('ascii', 'replace')
    at = 'on ' + str(time[0]) + ' ' + t_units.decode('ascii', 'replace')
    plt.title('{0}\n{1}\n{2}'.format(basename, desc, at))
    fig = plt.gcf()
    pngfile = "{0}.py.png".format(basename)
    fig.savefig(pngfile)
    

File: ICESat2/ATL15_AA_0314_01km_002_02_nc.m
%
%  This example code illustrates how to access and visualize NSIDC ICESat-2 
% ATL15 version 2 HDF5 file in MATLAB. 
%
%  If you have any questions, suggestions, comments on this example, please 
% use the HDF-EOS Forum (http://hdfeos.org/forums). 
%
%  If you would like to see an  example of any other NASA HDF/HDF-EOS data 
% product that is not listed in the HDF-EOS Comprehensive Examples page 
% (http://hdfeos.org/zoo), feel free to contact us at eoshelp@hdfgroup.org or 
% post it at the HDF-EOS Forum (http://hdfeos.org/forums).
%
% Usage:save this script and run (without .m at the end)
%                                   
% $matlab -nosplash -nodesktop -r ATL15_AA_0314_01km_002_02_nc
%
% Tested under: MATLAB R2023a
% Last updated: 2023-05-11

% Set the HDF5 File.
f = 'ATL15_AA_0314_01km_002_02.nc';

% Read the datasets.
ds = '/dhdt_lag1/dhdt';
data = h5read(f, ds);
data = squeeze(data(:,:,1));

% Read attributes.
fillvalue = h5readatt(f, ds, '_FillValue');
ln = h5readatt(f, ds, 'long_name');
units = h5readatt(f, ds, 'units');

% Process fill value.
data(data==fillvalue) = NaN;

x = h5read(f, '/dhdt_lag1/x');
y = h5read(f, '/dhdt_lag1/y');
t = h5read(f, '/dhdt_lag1/time');
units_t = h5readatt(f, '/dhdt_lag1/time', 'units');

% Transform location data.
proj = projcrs(3031);
[x_m, y_m] = meshgrid(y, x);
[lat, lon] = projinv(proj, x_m, y_m);

latlim = [floor(min(min(lat))), ceil(max(max(lat)))];

fig = figure('Name', f, 'Renderer', 'zbuffer', ...
             'Position', [0,0,800,600], 'visible', 'off');
pole=[-90 0 00];
axesm('MapProjection','stereo', ...
      'MapLatLimit', latlim, ...
      'Origin', pole ,'Frame','on','Grid','on', ...
      'MeridianLabel','on','ParallelLabel','on','MLabelParallel', ...
      'north')
% Align data with map - add 90 and fliplr.
lon = lon + 90;
surfm(lat, lon, fliplr(data));
colormap('Jet');

h = colorbar();
coast = load('coastlines.mat');         
plotm(coast.coastlat,coast.coastlon, 'k');
title({f;ln;['on ', num2str(t(1)), ' ', units_t]}, 'Interpreter', 'None', ...
      'FontSize', 12,'FontWeight','bold');

set(get(h, 'title'), 'string', units, ...
                  'FontSize', 12, 'FontWeight', 'bold');
saveas(fig,[f '.m.png']);

exit;

